code/xcvm/lib/core/src/shared.rs

```diff
@@ -15,3 +15,139 @@ pub fn encode_base64<T: Serialize>(x: &T) -> StdResult<String> {
 pub fn decode_base64<S: AsRef<str>, T: DeserializeOwned>(encoded: S) -> StdResult<T> {
 	from_binary::<T>(&Binary::from_base64(encoded.as_ref())?)
 }
+
+/// A wrapper around a type which is serde-serialised as a string.
+///
+/// For serde-serialisation to be implemented for the type `T` must implement
+/// `Display` and `FromStr` traits.
+///
+/// ```
+/// # use xc_core::Displayed;
+///
+/// #[derive(serde::Serialize, serde::Deserialize)]
+/// struct Foo {
+///     value: Displayed<u64>
+/// }
+///
+/// let encoded = serde_json_wasm::to_string(&Foo { value: Displayed(42) }).unwrap();
+/// assert_eq!(r#"{"value":"42"}"#, encoded);
+///
+/// let decoded = serde_json_wasm::from_str::<Foo>(r#"{"value":"42"}"#).unwrap();
+/// assert_eq!(Displayed(42), decoded.value);
+/// ```
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[derive(
+	Copy,
+	Clone,
+	Default,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Hash,
+	scale_info::TypeInfo,
+	derive_more::Deref,
+)]
+#[repr(transparent)]
+pub struct Displayed<T>(pub T);
+
+impl<T> parity_scale_codec::WrapperTypeEncode for Displayed<T> {}
+impl<T> parity_scale_codec::WrapperTypeDecode for Displayed<T> {
+	type Wrapped = T;
+}
+
+impl<T: core::fmt::Display> core::fmt::Display for Displayed<T> {
+	fn fmt(&self, fmtr: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+		core::fmt::Display::fmt(&self.0, fmtr)
+	}
+}
+
+impl<T: core::fmt::Display> core::fmt::Debug for Displayed<T> {
+	fn fmt(&self, fmtr: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+		core::fmt::Display::fmt(&self.0, fmtr)
+	}
+}
+
+impl<T: core::fmt::Display> serde::Serialize for Displayed<T> {
+	fn serialize<S: serde::Serializer>(&self, ser: S) -> Result<S::Ok, S::Error> {
+		ser.collect_str(&self.0)
+	}
+}
+
+impl<'de, T> serde::Deserialize<'de> for Displayed<T>
+where
+	T: core::str::FromStr,
+	<T as core::str::FromStr>::Err: core::fmt::Display,
+{
+	fn deserialize<D: serde::Deserializer<'de>>(de: D) -> Result<Self, D::Error> {
+		de.deserialize_str(DisplayedVisitor::<T>(Default::default()))
+	}
+}
+
+/// Serde Visitor helper for deserialising [`Displayed`] type.
+struct DisplayedVisitor<V>(core::marker::PhantomData<V>);
+
+impl<'de, T> serde::de::Visitor<'de> for DisplayedVisitor<T>
+where
+	T: core::str::FromStr,
+	<T as core::str::FromStr>::Err: core::fmt::Display,
+{
+	type Value = Displayed<T>;
+
+	fn expecting(&self, fmt: &mut core::fmt::Formatter) -> core::fmt::Result {
+		fmt.write_str("a string")
+	}
+
+	fn visit_str<E: serde::de::Error>(self, s: &str) -> Result<Self::Value, E> {
+		T::from_str(s).map(Displayed).map_err(E::custom)
+	}
+}
+
+impl<T> From<T> for Displayed<T> {
+	fn from(value: T) -> Self {
+		Displayed(value)
+	}
+}
+
+macro_rules! impl_conversions {
```

Comment:
`From` can handle one of the impls but I don’t see how the rest could be handled sadly.  Perhaps this could be simplified somehow with specialisation but that might end up being more complex than the current code.